cmd/snaplock/runinhibit/inhibit.go

```diff
@@ -283,3 +278,113 @@ func RemoveLockFile(snapName string) error {
 
 	return nil
 }
+
+// needed for better mocking
+type ticker interface {
+	Wait()
+	Stop()
+}
+
+type timeTicker struct {
+	*time.Ticker
+}
+
+func (t *timeTicker) Wait() {
+	if t.Ticker != nil {
+		<-t.Ticker.C
+	}
+}
+
+var newTicker = func(interval time.Duration) ticker {
+	return &timeTicker{time.NewTicker(interval)}
+}
+
+// WaitWhileInhibited blocks until snap is not inhibited anymore returning
+// a locked hint file lock.
+//
+// "notInhibted" and "inhibited" callbacks are called with the hint file lock held.
+// If inhibited callback returns true, WaitWhileInhibited returns instantly without
+// waiting for the snap not being inhibited.
+//
+// NOTE: It is the caller's responsibility to release the returned file lock.
+var WaitWhileInhibited = func(snapName string, notInhibited func() error, inhibited func(hint Hint, inhibitInfo *InhibitInfo) (cont bool, err error), interval time.Duration) (flock *osutil.FileLock, retErr error) {
+	ticker := newTicker(interval)
+	defer ticker.Stop()
+
+	if err := os.MkdirAll(InhibitDir, 0755); err != nil {
+		return nil, err
+	}
+
+	for {
+		flock, err := openHintFileLock(snapName)
+		if err != nil {
+			return nil, err
+		}
+		// release lock if we return early with an error
+		defer func() {
+			// keep lock if no errors occur
+			if retErr != nil {
```

Comment:
I don't think `retErr` is set elsewhere, but I think all the error cases in the for loop need to set it. Perhaps that would be an argument for making the returned `retErr` simply called `err`. Or could explicitly rename all the other ones so it's clear that they set the return error.

Perhaps you're right though, it'll assign whatever err is returned to `retErr`, and it's not necessary to do so explicitly. This is one of those scary Go things to me, but up to you.

Reply:
I am using a separate name intentionally here for readability, `err` will work though but I wanted to avoid the confusion about which error is read due to scopes.

Named return values are assigned before defer calls: https://go.dev/blog/defer-panic-and-recover
> 3. Deferred functions may **read** and assign to the returning function’s named return values.

Reply:
`TestWaitWhileInhibitedCallbackError` exercises this cleanup defer call.

Reply:
Very good point, thanks for explaining!